pkg/dialer/dialer.go

```diff
@@ -0,0 +1,93 @@
+// Copyright 2020 Google LLC
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     https://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dialer
+
+import (
+	"context"
+	"crypto/rand"
+	"crypto/rsa"
+	"net"
+	"sync"
+
+	"github.com/kurtisvg/cloud-sql-go-connector/internal/cloudsql"
+	sqladmin "google.golang.org/api/sqladmin/v1beta4"
+)
+
+var (
+	once sync.Once
+	d    *dialManager
+)
+
+// Dial returns a net.Conn connected to the Cloud SQL instance specified. The instance argument must be the
+// instance's connection name, which is in the format "project-name:region:instance-name".
+func Dial(ctx context.Context, instance string) (net.Conn, error) {
+	return getDialer().dial(ctx, instance)
+}
+
+// defaultDialer provides the singleton dialer as a default for dial functinons.
+func getDialer() *dialManager {
+	once.Do(func() {
+		// TODO: Do this better
+		key, err := rsa.GenerateKey(rand.Reader, 2048)
+		if err != nil {
+			panic(err)
+		}
+		client, err := sqladmin.NewService(context.Background())
+		if err != nil {
+			panic(err)
+		}
+		d = &dialManager{
+			lock:     &sync.RWMutex{},
+			imap:     make(map[string]*cloudsql.Instance),
+			sqladmin: client,
+			key:      key,
+		}
+
+	})
+	return d
+}
+
+type dialManager struct {
+	lock *sync.RWMutex
```

Comment:
The reason I made it a pointer was:
> A RWMutex must not be copied after first use.

Making it a part of the struct means that the dialManager struct must be passed by reference rather than value (which makes it hard to guarantee immutability).

Reply:
As long as we're passing a pointer to a `dialManager`, the mutex won't be copied. In fact, since the `dialManager` holds a map referencing connections, we'll probably always want to pass to a pointer to `dialManager`. If we wanted to share the lock itself, then for certain we would pass a pointer to it.

Reply:
> we're passing a pointer to a dialManager

Hmm - why would we want to return a pointer rather than the value? The map is already a reference type, so it seems safer passing by value?

Reply:
Discussed offline